### PR TITLE
systemd: 249 -> 250

### PR DESCRIPTION
**Pull request by jonringer**
###### Motivation for this change
Continuation of #150491

Opening PR so I can start large build job before getting on a plane.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by flokli** [comment]
This needs to target staging, and needs to be bumped to systemd/systemd-stable v250.3.

Furthermore, we should at least skim over the changes from v250..v250.3  and verify that they don't break any NixOS things.

**Comment by jonringer** [comment]
Yea, traveling right now. I'll mark as draft until it's actually ready

**Comment by jonringer** [comment]
I'll do a proper rebase and merge conflict analysis once I get home.

**Comment by jonringer** [comment]
@ofborg eval

File: nixos/doc/manual/release-notes/rl-2205.section.md
Comment (flokli) [inline nixos/doc/manual/release-notes/rl-2205.section.md:75]: This defaults flip been reverted in v250.3 (which this PR doesn't include yet).

@jonringer We should bump and revert this release note section.

File: pkgs/os-specific/linux/systemd/default.nix
Comment (flokli) [inline pkgs/os-specific/linux/systemd/default.nix:136]: Can we use `$SOURCE_DATE_EPOCH` instead?

File: nixos/doc/manual/release-notes/rl-2205.section.md
Comment (jonringer) [reply nixos/doc/manual/release-notes/rl-2205.section.md:75]: it does contain 250.3

I squashed all of the bumps together

Comment (jonringer) [reply nixos/doc/manual/release-notes/rl-2205.section.md:75]: I will revert the release notes

**Comment by Izorkin** [comment]
@jonringer commit `systemd: 249.5 -> 250.3` needed rename to `systemd: 249.7 -> 250.3`

File: nixos/doc/manual/release-notes/rl-2205.section.md
Comment (jonringer) [reply nixos/doc/manual/release-notes/rl-2205.section.md]: reverted

**Comment by jonringer** [comment]
rebased on latest master

File: pkgs/os-specific/linux/systemd/default.nix
Comment (flokli) [reply pkgs/os-specific/linux/systemd/default.nix:136]: It seems we can't, as we just fetch a git tag.  `SOURCE_DATE_EPOCH` is 315532800 (1980-01-01T00:00:00+00:00).

Comment (flokli) [reply pkgs/os-specific/linux/systemd/default.nix:136]: Let's remove the commented-out

```
# "-Dtime-epoch=1"
```

from L480 though.

**Comment by Mic92** [comment]
I saw your PR too late and opened up https://github.com/NixOS/nixpkgs/pull/156438

File: pkgs/os-specific/linux/systemd/default.nix
Comment (Mic92) [inline pkgs/os-specific/linux/systemd/default.nix:544]: I don't think we should attempt to patch tests.
I changed the code to replace tests instead: https://github.com/NixOS/nixpkgs/pull/156438/files#diff-a284772b505bd573f834b58591d380128545257d9e02601dbe112548526c8710R595-R597

File: pkgs/os-specific/linux/systemd/default.nix
Comment (Mic92) [inline pkgs/os-specific/linux/systemd/default.nix:418]: This option was just renamed and on staging we also use it:

```
"-Dlibcryptsetup=${lib.boolToString withCryptsetup}" 
```

**Comment by Mic92** [comment]
I am closing this just now to avoid pings on rebase...

**Comment by Mic92** [comment]
```
systemctl --version
systemd 250 (250.3)
+PAM +AUDIT -SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT -GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK -XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified
```

works on my laptop

File: pkgs/os-specific/linux/systemd/default.nix
Comment (flokli) [inline pkgs/os-specific/linux/systemd/default.nix:587]:
```suggestion
            echo "Not all references to '${search}' have been replaced. Found the following matches:"
```

File: pkgs/os-specific/linux/systemd/default.nix
Comment (flokli) [inline pkgs/os-specific/linux/systemd/default.nix:689]: 31e5b8dc2198eacec3e009e21fc8226486aa674e

**Review by flokli — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/systemd/default.nix
Comment (Mic92) [reply pkgs/os-specific/linux/systemd/default.nix]: added myself again.

**Comment by jonringer** [comment]
I'll try this on my laptop as well.

I think we could probably merge into staging after next staging-next cycle begins

**Comment by jonringer** [comment]
I'm running a hydra job on this: https://hydra.jonringer.us/jobset/nixpkgs/systemd250

**Comment by flokli** [comment]
> I think we could probably merge into staging after next staging-next cycle begins

@jonringer there have been some more staging cycles (https://github.com/NixOS/nixpkgs/pull/158080, https://github.com/NixOS/nixpkgs/pull/158080).

This also needs a rebase, as there have been some conflicts. Is there anything else preventing this from being merged? (I see the PR title still contains a "WIP" too).

**Comment by Mic92** [comment]
I am using this for some time on my laptop now without noticeable impact so far.

**Comment by jonringer** [comment]
@flokli I was going to use this staging cycle for glibc https://github.com/NixOS/nixpkgs/pull/133431#issuecomment-1053697179. Once it begins, I think we can merge this.

**Comment by jonringer** [comment]
I can still rebase though

**Review by ius — CHANGES_REQUESTED** [review]
Fails to build on EFI enabled RISC-V because the requested EFI linker (`efi-ld=gold`) is unsupported. According to Wikipedia gold only supports x86, x86-64, ARM, PowerPC, TileGX.

~~If there is no particular reason to pick gold over bfd, I suggest passing `-Defi-ld=bfd` instead.~~

Edit: I suggest just removing the option altogether, so it will figure out a suitable default by itself. See comment about changed semantics below as to why this is okay for v250.

File: pkgs/os-specific/linux/systemd/default.nix
Comment (flokli) [reply pkgs/os-specific/linux/systemd/default.nix:463]: Did this build before this PR?

Comment (ius) [reply pkgs/os-specific/linux/systemd/default.nix:463]: Yes. In systemd v249.7 the variable was used to invoke the linker directly (see [meson.build](https://github.com/systemd/systemd-stable/blob/v249.7/src/boot/efi/meson.build#L276)), so that's why `ld` was valid and defaulted to `bfd`.

In systemd v250.3 it's passed to `-fuse-ld` instead.

Comment (mweinelt) [reply pkgs/os-specific/linux/systemd/default.nix:463]: https://github.com/NixOS/nixpkgs/pull/150491#discussion_r767351535

**Comment by flokli** [comment]
I rebased this onto latest staging, fixed the conflicts, and dropped `-Defi-ld=gold` as suggested in https://github.com/NixOS/nixpkgs/pull/156096#discussion_r820131894.

Considering this has been open for quite some time, and tested by some people already, I'll run the `nixosTests.systemd{-boot}` tests once more, and then we should merge this.

**Comment by flokli** [comment]
`nixosTests.systemd-boot` succeeded.

`nixosTests.systemd` couldn't run, due to `/nix/store/i3axx4kkdbfl01r0dfsps9qnj8biryxh-python3.9-scikit-build-0.12.0.drv` being broken right now. However, this has been sufficiently tested previously.

Let's get this into staging.

**Comment by vcunat** [comment]
This merge broke `nixosTests.containers-ip`.  I checked the commit right before this merge and the merge commit (after some bisection steps, as I started with a range of many hundred commits).

Maybe also some other small-channel blockers from this list regressed together with it: https://hydra.nixos.org/eval/1748773#tabs-still-fail  (you can see logs there or reproduce cheaply on that particular commit)

**Comment by flokli** [comment]
Thanks for the heads-up, I'll take a look.

**Comment by flokli** [comment]
I built this locally. The container startup fails while trying to bind-mount `/nix/var/nix/daemon-socket` into the container:

```
machine # [    5.364468] systemd[1]: Starting Container 'webserver'...
machine # [    5.448067] container webserver[879]: Spawning container webserver on /var/lib/containers/webserver.
machine # [    5.448712] container webserver[879]: Press ^] three times within 1s to kill container.
machine # [    5.450500] container webserver[885]: Failed to stat /nix/var/nix/daemon-socket: No such file or directory
machine # [    5.454873] systemd[1]: container@webserver.service: Main process exited, code=exited, status=1/FAILURE
machine # [    5.455510] systemd[1]: container@webserver.service: Failed with result 'exit-code'.
machine # [    5.456098] systemd[1]: Failed to start Container 'webserver'.
machine # [    5.763319] systemd[1]: container@webserver.service: Scheduled restart job, restart counter is at 2.
machine # [    5.769093] systemd[1]: Stopped Container 'webserver'.
machine # [    5.770123] systemd[1]: Starting Container 'webserver'...
```

I ran the interactive test driver, and looked at the VM. Indeed `/nix/var/nix/daemon-socket` doesn't exist, because `nix-daemon.socket` failed to activate:

```
[  237.187747] systemd[1]: Nix Daemon Socket was skipped because of a failed condition check (ConditionPathIsReadWrite=/nix/var/nix/daemon-socket).
```

**Comment by flokli** [comment]
Okay, I figured this out - this is actually a bug in the upstream nix-daemon.socket unit.

I have a workaround in nixpkgs, and will send a fix to the Nix repo.

**Comment by flokli** [comment]
@vcunat the workaround is in https://github.com/NixOS/nixpkgs/pull/164644, and the proper fix in Nix is in https://github.com/NixOS/nix/pull/6282.

**Pull request merged**
